Rewrite with Slice First Design

RishabhRD opened the pull request:
# Description
This is a long term experimental PR. This is a rewrite for the whole project with intentions to solve tech debts the project have currently. 

**NOTE**: This PR would be the base discussion point for any core changes. There would be more PRs in future to port already implemented algorithms to this design.

## Problems with current design
1. Multiple variants of same algorithm for composition purposes (pass positions explicitly in algo:: version). #16
2. InputRange is not a good model for single pass range. #13
3. Algorithms that require elements in memory are not expressible. #12
4. Semantics of View is unclear. #18 
5. Mutability problems with views and ranges are combined. #14 

See [old design docs](https://github.com/RishabhRD/rs-stl/tree/deprecated_view_code/docs) for knowing more.

## New design

### Highlights of new design
1. Slices are first class entity. See #16 for knowing more.
2. There would be no support for single pass ranges. Iterator is a good model for the same which solves problem 2.
3. BoundedRange is dropped because with multipass ranges, almost every range have a bound. And it was adding complexity for slice first model.
4. View is dropped as it didn't have clear semantics which solves problem 4.
5. Clear distinction if element of range is present in memory or not.
6. Drop variant of algorithms that needs passing explicit positions for subrange notation.

### Details

**NOTE:** I am still trying to figure out mutation model for this new design. I have put my initial naive thoughts currently.

![rs-stl-base](https://github.com/user-attachments/assets/5760af56-d005-43be-8ccc-e3119c22ecdd)

- `Range` is a multipass linear sequence of elements. Range defines an associated type `SubRange` to denote the type of Slice of range. This goes into direction of slice-first-design suggested by @dabrahams in #16 . Here the key is restriction of slice of slice is slice itself. This enables us to slice the given range any number of times without loosing efficiency that is common in algorithms like quick-sort. This would solve problem 1 as algorithm variant with explicit positions is now not needed.
- `Collection` defines range whose elements are present in memory. This is expressed by `Collection` trait providing `at` method that returns reference to element.
- `LazyCollection` defines range whose elements are computed lazily on element access. This is expressed by `LazyCollection` trait providing `at` method that returns value of element.
- This clear distinction of `Collection` and `LazyCollection` would solve problem 3. One of ambitious goal would be to write algorithm for the problem "**sort every word of a given string**" using composition of `split_lazy`, `sort_range` and `for_each` algorithms.
- `Collection` and `LazyCollection` also provides `iter` method to iterate over reference to elements and elements respectively.
- `ReorderableRange` provides reordering capability to range. `Collection` and `LazyCollection` both might support the same if semantics of reordering is well defined. `ReorderableRange` provides a `SubRangeMut` type alias to work with mutable slices.
- `MutableCollection` provides ability to mutate element by providing mutable access to elements. `LazyCollection` can't provide mutability because their elements don't exist in memory.

### Implementation Problems right now
- To achieve slice first design, `SubRange` associated type should be defined, however, `SubRange` have semantics of lifetime associated with it too, so need to use GAT for the same. Extending it to traits like `BidirectionalRange` requires use of HRTB. However, mixing HRTB with GAT leads to [borrow checker limitation](https://github.com/rust-lang/rust/issues/87479) currently and it interprets lifetime of `SubRange` as `'static`.
- For the workaround, range is broken into 2 more parts, `RangeBase` and `SubRangeable` with involving trick around implicit bound inference. This looks ugly but works.
- Mutable ranges are not yet implemented as I am fighting with borrow checker with above limitations too. Lifetime of mutable subranges are even hard to get right as rust can't have 2 mutable references to a variable of same lifetime at a time. Thus even `SubRangeMut` of `SubRangeMut` would not exactly be `SubRangeMut` (lifetime would be different), so I need to think how to express this whole thing with borrow checker limitations plus obvious fight.
- Currently, it is needed to repeat `SubRange` bounds anywhere we use `Collection` or `BidirectionalRange`, etc. Example:
```rust
impl<R> Collection for Slice<'_, R>
where
    R: Collection,
    for<'a> <R as SubRangeable<'a>>::SubRange: Collection, // <--- This is ugly
{...}
```
This looks ugly to use, but I am not sure how to avoid it.

#### My thoughts
Current design actually seems good, but the implementation in rust makes it little ugly and hard to achieve. Rust borrow checker limitations and complex borrow checking rules make implementation a tedious process.

dabrahams commented:
I'm not sure what your question about the mutation model is.

Rust suffers from not having [subscripts with inversion of control](https://forums.swift.org/t/modify-accessors/31872), which probably forces your distinction between lazy and non-lazy collections.  But you could potentially unify them in Rust by making the common API take a lambda that receives the element.  In Swift:

```swift
protocol Collected { // ignore the name
  associatedtype Position
  associatedtype Element
  func read<R>(_ x: Position, (borrowing Element)->R) -> R
}

protocol MutableCollected: Collected {
  associatedtype Position
  associatedtype Element
  func write<R>(_ x: Position, (inout Element)->R) -> R
}
```

Also nice: you can avoid lifetime annotations by using these APIs.

RishabhRD commented:
@dabrahams regarding mutation model I am afraid that the current subrange design can't be implemented in rust with the borrow checker limitations and bugs it has right now. In terms of code, this should work perfectly fine:
```rust
pub trait MutableRange: Range {
    type SubRangeMut<'a>: MutableRange<
        Element = Self::Element,
        Position = Self::Position,
        SubRangeMut<'a> = Self::SubRangeMut<'a>,
        SubRange<'a> = Self::SubRange<'a>,
    >
    where
        Self: 'a;

    fn slice_mut(
        &mut self,
        from: Self::Position,
        to: Self::Position,
    ) -> Self::SubRangeMut<'_>;
}
```
However, composing this with abstractions like bidirectional range makes SubRangeMut mandatory to have `'static` lifetime, that makes the whole thing unusable.
And even the workaround that I did for immutable subrange, would not work here as it mandates `SubRangeMut` to have same lifetime as `Self`, however, that is not possible with mutable references in rust as per borrow rules.

After long thought, I am thinking to settle on following idea:
- Ranges would not know about subrange type.
- Instead, there would be 2 generic subrange classes: `Slice` and `MutableSlice` that stores reference to range, start position and end position.
- `Slice` and `MutableSlice` would maintain the responsibility of "slice of a slice should be slice itself" by providing appropriate methods.
- `Slice` and `MutableSlice` would themselves be ranges.
- All the algorithms in library would work on slice, rather than working on raw ranges. (e.g., find_if would accept `Slice<Range>` as argument)
- Provide type coercion for raw range types so that `&range` and `&mut range` would be converted to Slice and MutableSlice implicitly when passed to function accepting Slice or MutableSlice. This is for making the library user-friendly.
- This would be very similar behaviour what rust has for inbuilt slice type and array types. Just that in rust, inbuilt slice doesn't share same position as the range by default.

Now coming to LazyCollection, I got to know about this idea of inversion of control on subscript from your cpponsea talk.

Regarding my thought process on LazyCollection (this might be total over-engineering, if so please guide me here), I was thinking of following usecase:
```rust
fn sort_every_word<R: MutableRange>(rng: MutableSlice<R>){  // ignore the syntax for now
   let words = rng.split(' ');
   let start = words.start();
   let end = words.end();
   while start != end {
        let mut word = words.at_mut(&start);
        word.sort(); // <--- this makes sense
        // word = some_other_slice; <--- does not make sense
        start = words.after(start);
   }
}
```
In the example, `words` have element type of `MutableSlice`. `words.at_mut(&i)` should return the slice by value not by mutable reference as `word.sort()` makes sense, however operation like `word = some_other_slice` doesn't make any sense. This thing is different from collections where `arr.at_mut(&i) = e` makes sense.

I think reason for this kind of difference is because `Collection`'s `Range::Element` is stored in memory so mutating it freely makes sense. But for lazy collections, `Range::Element` is not stored in memory (e.g., the `word` slice in `split` example was not actually stored in `words` range, however, it was created on element access) for which mutating elements freely doesn't make sense. So, lazy collections might provide a controlled mutable view of underlying elements that would allow mutating possible underlying collection wrapped inside lazy collection in some senseful way.

RishabhRD commented:
@dabrahams this is an example of quick-sort implementation with the approach I described where algorithm accepts slice as argument:
```rust
fn partition_on<R: ReorderableRange, Pred>(
    mut slice: MutableSlice<R>,
    i: &R::Position,
    pred: Pred,
) -> R::Position
where
    Pred: Fn(&R::Element, &R::Element) -> bool,
{
    todo!()
}

fn sort<R: ReorderableRange + RandomAccessRange>(mut slice: MutableSlice<R>)
where
    R::Element: Ord,
{
    let start = slice.start();
    let mut end = slice.end();
    if start == end {
        return;
    }
    end = slice.before(end);
    let p = partition_on(slice.clone(), &end, |x, end_element| x < end_element);
    sort(slice.clone().prefix_mut(p.clone()));
    sort(slice.suffix_mut(slice.after(p)));
}

fn main(){
  let mut arr = [2, 1, 3];
  sort(arr.slice_mut()); // This looks ugly :(
}
```
Currently one can write this kind of code with the above approach. I am unsure how feasible is the coercion thing, so currently user needs to pass `arr.slice_mut()` for working with algorithm that looks so verbose to use.

One approach to avoid this verbosity without type coercion is to:
- take slice as argument in algorithms like sort which needs to operate on subrange.
- take reference to range as argument in algorithms like find_if that doesn't deal with subrange at all.

However, this would create asymmetry that also doesn't feel right.

One thought is to re-evaluate the conclusion that there is no workaround possible for borrow checker bug in rust for mutable slices currently as I might miss it because I am new to rust language.

dabrahams commented:
> regarding mutation model I am afraid that the current subrange design can't be implemented in rust with the borrow checker limitations and bugs it has right now.

I am pretty good at working around compiler bugs (I had to do a lot of that in the early days of Swift).  If you could check in the code that's supposed to work somewhere, I could try to make it compile.

I have a lot to read here; I'll be responding to what you wrote piecemeal.

dabrahams commented:
I'm not sure your sorting use case makes sense.  Won't these words have different lengths?  Swapping their elements, if you could do it at all, would cause an O(N) shift of the intermediate elements.  Am I missing something?

dabrahams commented:
I wonder about the name `LazyCollection`.  Should a collection that merely projects references to stored parts of an underlying collection's elements be considered lazy?  The elements are stored after all, but retrieving them costs a computation.  Where does that fit into your concept hierarchy?

dabrahams commented:
I don't know what Rust culture says about doing things this way, but in Swift we would always inject the algorithms as methods.  That has huge usability benefits over free functions (reduced nesting, code completion).

dabrahams commented:
The pattern where slices are passed by transfer-of-ownership has some real downsides; these slices expose themselves as references to the underlying data, which is why you have to `clone()` them in quicksort.  In the Swift/Hylo model (which may be problematic for Rust  for all I know), the slice is treated as a value (of course) that fully owns its elements.  So in Rust terms, `quicksort` would accept an `&mut` slice.

I wonder if MutableCollection and LazyCollection should really be distinct.  The could obviously be unified by making the return type of the indexing operation be an associated type.

dabrahams commented on src/core.rs:
```diff
@@ -1,225 +1,215 @@
+/// Models a lazy collection which provides access to mutable view of element at any position.
+pub trait MutableLazyCollection: MutableRange + LazyCollection {
+    /// Mutable view of element.
+    type ElementMut;
```
It's a bit suspect that there are no constraints on this type and that there is no apparent underlying element type.  I am concerned that your traits might not be meaningful abstractions.

It looks a bit like you are inventing concepts here rather than discovering them.  Generic programming proceeds from efficient, _non-generic_ algorithm implementations.  So for example, take some generalized algorithm (let's say a linear search, for simplicity), and write down a non-generic implementation for all the examples of data structures you want to operate on, including the lazy ones.  Don't worry about unifying their APIs.  Then lift the generic implementation and implementation requirements out of those concrete implementations.  Then you cluster requirements into concepts.  See [this summary](https://stackoverflow.com/a/72342369/125349) and read "From Mathematics to Generic Programming"

RishabhRD replied:
I fully agree with you on this. I literally hate MutableLazyCollection trait as it deviates heavily from concrete code examples we write and doesn't seem to follow the idea of generic programming.

But here is the process what led me to write this trait:

I started from concrete example of sorting every individual words in string. To be clear: "bac ba" -> "abc ab".
Now, for that I would do something like: `let words = str.split(' ');`. Here the element type of `words` range would be `Slice<Array>`, where slice is:
```rust
pub struct Slice<'a, R: Range> {
    m_range: &'a R,
    m_start: R::Position,
    m_end: R::Position,
}
```
Slice contains reference to original range.

Now for sorting first word of `words` I need to call `at_mut` on start position. But if `at_mut` returns `&mut Slice<Array>` that would be of no use for me at all as Slice has immutable reference to original range and thus I can't mutate range.
Thus I need something like `MutableSlice<Array>` that contains mutable reference to given range and that led to write a `ElementMut` associated type. I was unable to figure out any good relationship between Element and ElementMut and thus I left it unconstrained in the hope someday I would be able to figure this out.

This is also a reason for me to think if "sorting every word in a string" is a good problem for influencing the design of library because certainly it is adding non-intuitive complications.

I guess Hylo doesn't suffer from this as Slices in hylo are mutable or immutable based on the given collection is mutable or immutable. I might be wrong here.

Regarding "From Mathematics to Generic Programming.", I started it but didn't complete it. I should definitely complete it now as there would be no better time for this.

dabrahams replied:
I think I understand your story about slices… except for the part about how that led you to create `MutableLazyCollection`.  You say that it led you there, but not how/why.

> I guess Hylo doesn't suffer from this as Slices in hylo are mutable or immutable based on the given collection is mutable or immutable. I might be wrong here.

I think we haven't entirely settled that question.  To do that we would have to cast the remote part that represents the underlying collection to mutable so that there would be a single representation, but it can't be misused of course because you would need to mutate the slice to mutate its (remote) parts. You lose the ability to pop elements off the ends of the slice, although I think we lose that anyway, as there's no way to project a mutable thing out of an immutable one.  This fact makes slice-first in Hylo look slightly less attractive, to be honest.  You can always project a new, shorter, slice out of the old one.

RishabhRD replied:
> I would guess that there would at least be an `into` relationship between these types, if you have to have them.

Yes!!! ElementMut should be convertible to Element by some function F. I guess this fact might be used as a constraint here somehow. Let me try something on this.
Also I think `Element` and `ElementMut` should accept some lifetime parameters if we wish to unify `Collection` and `LazyCollection` , `MutableCollection` and `MutableLazyCollection`.

RishabhRD commented:
> I am pretty good at working around compiler bugs (I had to do a lot of that in the early days of Swift). If you could check in the code that's supposed to work somewhere, I could try to make it compile.

@dabrahams This would be a really great help. First of all a really  big thanks for this as well as for guiding me on this project.
I have put the code that ideally should work in [ideal_slice_code](https://github.com/RishabhRD/rs-stl/tree/ideal_slice_code) branch. I have removed everything but the essentials. `src/array_slice.rs` doesn't compile but I guess it should ideally. Let me know if you need any other code changes or something in branch.

That branch code currently contains abstractions that you might not agree with. We will discuss about those points and I will change the branch code based on that.

Now I will try to reply to your comments in next comments.

RishabhRD commented:
> I'm not sure your sorting use case makes sense. Won't these words have different lengths? Swapping their elements, if you could do it at all, would cause an O(N) shift of the intermediate elements. Am I missing something?

I guess I was not clear in my use-case wording. Let me phrase it with an example:

Let's say I have a string: `"bac ba"`. I want to sort every word of string.

So, resultant string should be: `"abc ab"`.
and NOT: `"ba bac"`.

I mean individual words should be sorted.

I am not sure this is a good problem for influencing design of library but I thought it might be good if we can do this with composition of `split` and applying `sort` using `for_each`. Let me know if you were talking about this only... then I am missing something for sure.

This is also an example of range where individual elements (in this case slice) can be mutated but not reordered (2 slices can't be swapped).

RishabhRD commented:
Let me address `LazyCollection` abstraction. Personally I also don't like the idea of LazyCollection as a separate entity as I am aware that it would lead to combinatoric problems when Collections and LazyCollections will be used in mix like `zip` algorithm.

> Should a collection that merely projects references to stored parts of an underlying collection's elements be considered lazy? The elements are stored after all, but retrieving them costs a computation. Where does that fit into your concept hierarchy?

Currently I call it a LazyCollection only, but I know this is not a good term for the purpose. And thus I am also trying to retire LazyCollection.

So, let's say I only have `Collection` abstraction  and I will only return/read (inversion of control) reference to elements. Then I am afraid it would lead to loss of efficiency for these 2 usecases:
1. `lazy_arr.copy_to(new_arr)`. Here if `at` method returns a reference, then an additional copy would be needed for putting it into new_arr. On the other side if `at` method returns value (that is default behavior for lazy collection) then that copy would not be needed.
2. `arr.map(...).map(...)` suffers from the same problem of additional copying. (let's assume map has lazy behavior)

But for this I guess you suggested that still we can unify them using an associated type for `at` return type. I am not sure how much you like this kind of unification or do you think we should return only references from LazyCollection (i.e., discard LazyCollection abstraction and use inversion of control/proxy references).

Btw current master branch is actually using proxy references and collections and lazy collections are unified and is called Range. Just that sometimes, it lead me to hit the same borrow checker bug I described earlier on writing algorithms like monadic join.

I am curious to know what do you think about above usecases or I am overthinking/blundering something here.

RishabhRD commented:
> I don't know what Rust culture says about doing things this way, but in Swift we would always inject the algorithms as methods. That has huge usability benefits over free functions (reduced nesting, code completion).

I also like algorithms as methods only. I think rust culture is the same. (But I am surely not the right person to comment on that as I started learning rust 3 months back only).

I presented quick sort like free function as it was dependent on slice thing and writing extension trait for the same would have required good amount additional boilerplate that would have been a distraction for example purpose.

RishabhRD commented:
> It might be worth experimenting with [Generator](https://dev-doc.rust-lang.org/beta/std/ops/trait.Generator.html)s to see if you can achieve inversion of control that way.

This looks promising, I would experiment with this.

Regarding passing lambda to indexing function, I am unsure if that would affect ergonomics of API. I lack years of experience to take decision on this.

dabrahams commented:
> > It might be worth experimenting with [Generator](https://dev-doc.rust-lang.org/beta/std/ops/trait.Generator.html)s to see if you can achieve inversion of control that way.
> 
> This looks promising, I would experiment with this.

Looks like it's called coroutines now, and having looked more closely I'm not at all sure it can help.  The closure literal syntax is limiting, and it looks like you can't yield a reference to a local.  But if you can explain what _would_ be needed [here](https://github.com/rust-lang/rust/issues/43122#issuecomment-2631815995), that would be good.

> Regarding passing lambda to indexing function, I am unsure if that would affect ergonomics of API. I lack years of experience to take decision on this.

Oh, it _definitely_ would affect the ergonomics.  But the other option isn't great either.  Might be worth experimenting to see how it turns out.

dabrahams commented:
Your points about additional copies are well-taken.  There are lots of things that Swift, with its O(1) copies (and when I was working on it, universal copyability) makes a non-issue, and I tend to lose sight of those.  We need to think more carefully about these rvalue cases.  I wonder if you'd like to join us in [Hylo slack's](https://join.slack.com/t/val-qs97696/shared_invite/zt-2z9v21xgh-~N7ti2axbGzgRZfCFpUjXA) #stdlib channel to discuss some of these things?

RishabhRD commented:
> > This is also an example of range where individual elements (in this case slice) can be mutated but not reordered (2 slices can't be swapped).
> 
> I could be mistaken but I think that's fundamentally wrong, because in Rust everything can be moved, and Rust's `swap` is unconstrained. And I think that dictates the semantics of these swaps: they don't reorder anything in the underlying collection.

What I meant to say here is allowing swap_at operation in this use-case of `str.split(' ')` doesn't have clear semantics. Are you suggesting it should be supported and implemented as a no-op?

RishabhRD commented:
> Your points about additional copies are well-taken. There are lots of things that Swift, with its O(1) copies (and when I was working on it, universal copyability) makes a non-issue, and I tend to lose sight of those. We need to think more carefully about these rvalue cases. I wonder if you'd like to join us in [Hylo slack's](https://join.slack.com/t/val-qs97696/shared_invite/zt-2z9v21xgh-~N7ti2axbGzgRZfCFpUjXA) #stdlib channel to discuss some of these things?

Sure @dabrahams . Infact, I have already joined the hylo slack channel. Let's discuss the things there.

dabrahams commented:
> What I meant to say here is allowing swap_at operation in this use-case of str.split(' ') doesn't have clear semantics. Are you suggesting it should be supported and implemented as a no-op?

No, I'm saying it must be supported if the result is a mutable collection, and its semantics are to change the order of the slices in the result, but not in the underlying string. So (pardon my rust):

```rust
let mut x = "a b";
let mut y = x.split(' ');
y.swap_at(0, 1); // assuming positions are consecutive ints starting at 0
y[0][0] = 'c'; // assuming you can address the zeroth character of a string s with s[0]
```

Now `x` is `"a c"` and `y` is, roughly `["c", "a"]`

RishabhRD commented:
> No, I'm saying it must be supported if the result is a mutable collection, and its semantics are to change the order of the slices in the result, but not in the underlying string. So (pardon my rust):
> 
> ```rust
> let mut x = "a b";
> let mut y = x.split(' ');
> y.swap_at(0, 1); // assuming positions are consecutive ints starting at 0
> y[0][0] = 'c'; // assuming you can address the zeroth character of a string s with s[0]
> ```
> 
> Now `x` is `"a c"` and `y` is, roughly `["c", "a"]`

@dabrahams I don't see a way to implement this without any additional overheads. Just to be clear, `x.split(' ')` doesn't do any additional allocation at all. For split range, Position type is pair of base's positions. Using this position type, element (i.e, slice) can be computed on element access without any additional allocation.
Now for swapping elements of 2 positions for split, I am unsure if we can do the same without maintaining info about swapped elements in additional memory. Or am I missing some point here?

dabrahams commented:
>  Just to be clear, `x.split(' ')` doesn't do any additional allocation at all. For split range, Position type is pair of base's positions. 

Oh, I see, it acts something like a filter view.  Personally I would never have tried to make such a thing be a mutable collection.  Unless I'm missing something, it _is_ a fundamental law in Rust that all things that can be directly mutated can be swapped, so even if you don't provide `swap_at` directly, someone could 

```
let mut v = "my string".split(' ');
let mut t = " ".split(' ');
mem::swap(t.first(), &v.first());
mem::swap(v.last(), t.first());
```

This code will have some semantics and if it doesn't make sense, then your split view type should probably not be a mutable collection.

RishabhRD commented:
> Personally I would never have tried to make such a thing be a mutable collection.

@dabrahams I understand your take here. Making this kind of ranges mutable collection might lead to surprising mutations. My idea for making it mutable collection was just  to experiment how far we can take inplace mutations with generic programming in general. Do you think this is the limit?

> Unless I'm missing something, it _is_ a fundamental law in Rust that all things that can be directly mutated can be swapped, so even if you don't provide `swap_at` directly, someone could
> 
> ```
> let mut v = "my string".split(' ');
> let mut t = "_".split(' ');
> mem::swap(t.first(), v.first());
> mem::swap(v.last(), t.first());
> ```
> 
> This code will have some semantics and if it doesn't make sense, then your split view type should probably not be a mutable collection.

Regarding this code, This won't be technically allowed in rust. 

First this doesn't work with current design as currently we return slices by value (as slices are not actually in memory). However I understand that returning slices by value are problematic because the whole idea of value semantics is that **values are independent**. So in spirit of that, inversion of control should be adopted that we agree upon.

But even with inversion of control, swapping mutable slices would not be allowed by borrow checker. See ([Godbolt Link](https://godbolt.org/z/TeGe4rGvh)):
```rust
struct Array {
    pub arr: [i32; 5],
}

struct Slice<'a> {
    data: &'a mut [i32; 5],
    start: usize,
    end: usize,
}

impl Array {
    pub fn get_slice<'a, F>(&'a mut self, start: usize, end: usize, mut f: F)
    where
        F: FnMut(&mut Slice<'a>),
    {
        let mut slice = Slice {
            data: &mut (self.arr),
            start,
            end,
        };
        f(&mut slice);
    }
}

fn main() {
    let mut x1 = Array {
        arr: [1, 2, 3, 4, 5],
    };
    let mut x2 = Array {
        arr: [11, 12, 13, 14, 15],
    };

    x1.get_slice(0, 3, |x1_slice| {
        x2.get_slice(1, 2, |x2_slice| {
            std::mem::swap(x1_slice, x2_slice);
        });
    });
}
```
Not sure if rust is overly restrictive here. Will hylo allow that?

RishabhRD commented:
One more thing is, even if `str.split(' ')` creates  new array, then also that array would only be reorderable not fully mutable (like we expect for our usecase). Because elements of resultant array would `Slice` not `MutableSlice`. Another option is to provide `split` as well as `split_mut` method, but that looks bad.

dabrahams commented:
Over in Hylo land we are beginning to wonder if the use-cases for lazy multipass sequences are strong enough to warrant their inclusion; we're wondering if they should just be iterators.

dabrahams commented:
> @dabrahams This would be a really great help. First of all a really big thanks for this as well as for guiding me on this project.
I have put the code that ideally should work in [ideal_slice_code](https://github.com/RishabhRD/rs-stl/tree/ideal_slice_code) branch. I have removed everything but the essentials. src/array_slice.rs doesn't compile but I guess it should ideally. Let me know if you need any other code changes or something in branch.

The whole thing compiles just fine on my machine.  Am I missing something?

RishabhRD commented:
> The whole thing compiles just fine on my machine. Am I missing something?

@dabrahams I accidentally pushed some changes to this branch yesterday, my fault :sweat_smile: . Can you please take the pull again, where I added the code that doesn't compile but should compile.

dabrahams commented:
I looked; the problem is familiar to me: Rust can't deal with recursive constraints (e.g. where `Collection`'s `Slice` type must satisfy `Collection`); you should probably open an issue about that.  Swift used to have this problem and I used to hack around it in ways that leaked out into the user experience, but eventually the problems with that caused the compiler engineers to fix it… enough to compile most programs.  It was only really addressed in a principled way by Slava Pestov's requirement machine.

Then when we were exploring designs for Hylo we asked ourselves whether we could live without recursive constraints, and I built [this prototype](https://gist.github.com/dabrahams/bc9497acdc20001cde14ed57bc4b4965#file-norecursiveconstraints2-swift) to show that it was possible.  You ought to be able to use this pattern.

RishabhRD commented:
Hi @dabrahams I went through the link you provided. I have been doing similar stuff on [another branch](https://github.com/RishabhRD/rs-stl/tree/impl_workaround_slice_first) for workaround. However, then this problem converts to lifetime annotation problem. Even with dropping all the constraints on slice, I can't write the slice method for collection as I face lifetime related issues (I can't pass slice as argument to callback function, as it needs to satisfy some lifetime annotation requirements that I am unable to get my head around).

RishabhRD commented:
Hi @dabrahams, I am really excited to tell that finally I was able to make the slice first design work, thanks to the workaround you provided and **ofcourse to you**.

Recently I realised that the `SliceCore` in your workaround doesn't need to be Generic Associated Type in rust, rather it can just be a normal associated type without any need of lifetime parameter. And this observation helped me escape all those lifetime related limitations of borrow checker. I don't know how did I miss this thing for this long time. I have pushed the changes in this PR itself.

Current change is not an exact translation of the whole workaround. I took the idea of `Whole` thing only as I felt rust was able to handle the need for `CollectionCore` and `SliceCore`. I might miss something in this, please correct me if I am wrong here.

This is an implementation of quick sort with current design that compiles well:
```rust
pub trait RandomAccessReorderableCollectionExt:
    RandomAccessCollection + ReorderableCollection
where
    Self::Whole: RandomAccessCollection + ReorderableCollection,
{
    fn quick_sort(&mut self)
    where
        Self::Element: Ord,
    {
        let start = self.start();
        let mut end = self.end();
        if start == end {
            return;
        }
        end = self.before(end);
        let p = self.partition_on_pos(&end, |x, y| x < y);
        self.prefix_mut(p.clone()).quick_sort();
        self.suffix_mut(p).quick_sort();
    }
}
```
In this implementation `Self::Whole: RandomAccessCollection + ReorderableCollection,` this line is the only thing I felt could be redundant if rust supported a specific syntax, otherwise seems quite nice to me.

Regarding the other approach I was trying with lifetimes and GATs involved, many rust people told that rust is actually really terrible in handling that specific kind of lifetime calculus so, that approach is not possible in rust (atleast for the time being).

But I am glad we can do the whole thing without lifetime GATs as that would have just made library more complex most probably (due to rust borrow checker limitations).

Still I have one todo in core for handling that `LazyCollection` usecase, specifically unifying it with Collection. I need to decide whether we should have `Element` associated type that represents `type of values inside collection` or we need to have `Element` and `ElementMut` associated types in which `Element` would be like `reference to values inside collection` and `ElementMut` would be `mutable reference to values inside collection`.

dabrahams commented:
Very cool; congratulations!  FYI, you can implement `quick_sort` for `Collection`; you don't need random access.  I couldn't find partition_on_pos but I guess I can intuit what it's doing.  Use the first element instead of the last to avoid needing bidirectional access.

I think you'd better change the comment on `Whole`; the current one seems to indicate it should be called `SliceCore`.

> I felt rust was able to handle the need for CollectionCore and SliceCore. 

Absolutely, it seems you don't need to completely avoid recursive constraints on associated types.  The fewer workarounds the better.

Your `quick_sort` code is pretty gorgeous!  How does it compare with the current way you'd do it in Rust?

> In this implementation Self::Whole: RandomAccessCollection + ReorderableCollection, this line is the only thing I felt could be redundant

Can you not build the `Self::Whole` constraints into the definitions of `RandomAccessCollection` and `ReorderableCollection`?  Oh, bizarre: you did, but Rust doesn't seem to know it!  What does a specific syntax have to do with it?  It should just work without that line AFAICT.

I don't know why `prefix_mut` and `suffix_mut` need to consume their argument, maybe that's just a Rust style default for things that are likely to be small?  In Hylo the rule of thumb is pass by value (immutable reference in Rust) unless the value is going to escape (e.g. be stored or returned).

Would you like to discuss naming?  I don't really know how it's usually done in Rust, but I might have some useful input.

Regarding lazy collections, I think you ought to declare victory on this PR and come back to lazy collections in a second pass.  There are many complicated issues in this area.  @kyouko-taiga has some radical thoughts about this and I'd love to see what comes of a discussion.  That said, the more I come to understand Rust references the more I realize they are more like non-nullable pointers than like C++ or even Swift references.  So I think I would always use the first definition for `Element` and pick different names for those “reference to…” types.

RishabhRD commented:
> FYI, you can implement `quick_sort` for `Collection`; you don't need random access.

Yes, just realised that and this thing makes it even more beautiful. 
```rust
pub trait ReorderableCollectionExt: ReorderableCollection
where
    Self::Whole: ReorderableCollection,
{
    fn quick_sort(&mut self)
    where
        Self::Element: Ord,
    {
        if self.start() != self.end() {
            let p = self.partition_on_pos(&self.start(), |x, y| x < y);
            self.prefix_mut(p.clone()).quick_sort();
            self.suffix_mut(self.after(p)).quick_sort();
        }
    }
}
```
Infact, this is already way better than how rust handles sorting right now.

> I couldn't find partition_on_pos but I guess I can intuit what it's doing.

Yes, it doesn't exist right now. Will implement it while writing partitioning algorithms in future.

> Your `quick_sort` code is pretty gorgeous! How does it compare with the current way you'd do it in Rust?

Current rust way also looks quite nice in general but has some really deep problems. This is a sample implementation (implementing it as function not as method for readability as it would introduce good amount of boilerplate):
```rust
fn quicksort<T: Ord>(arr: &mut [T]) {
    if arr.len() > 1 {
        let pivot_index = partition_on_pos(arr, 0, |x, y| x < y);
        quicksort(&mut arr[0..pivot_index]);
        quicksort(&mut arr[pivot_index + 1..]);
    }
}
```
So, the problems with rust implementation are (and this holds for rust stdlib sort implementation too):
1. The implementation is actually there for `inbuilt slice` type of rust. Data structures like array and vector are implicitly convertible to inbuilt slice type by language feature named `Unsize Coercion`. Thus this implementation actually depends on language features for its usefulness.
2. The worse part is unsize coercion isn't achievable for user defined types and requires compiler magic (as implicit conversion should be limited for its harm). Thus a custom data structure can't be used with this quick sort implementation. There is something called `Deref Coercion` that can be used but again your data structure should be then just wrapper over array or vector or unsafe pointers.
3. The inbuilt slice type actually requires data should be contiguous in memory that itself is limiting as you pointed quick sort doesn't require even random access.

Will be answering other points in different comments.

RishabhRD commented:
> I think you'd better change the comment on `Whole`; the current one seems to indicate it should be called `SliceCore`.

Yep fixed that.

> Can you not build the `Self::Whole` constraints into the definitions of `RandomAccessCollection` and `ReorderableCollection`? Oh, bizarre: you did, but Rust doesn't seem to know it! What does a specific syntax have to do with it? It should just work without that line AFAICT.

This is the specific syntax I am talking about:
```rust
pub trait RandomAccessCollection:
    BidirectionalCollection<Position: Regular + Ord, Whole: RandomAccessCollection>
{
}
```
Rust supports it, just that it can't be recursive in nature like we have for Whole.

Ord constraint on Position would clarify this. Without this syntax it would be like:
```rust
pub trait RandomAccessCollection: BidirectionalCollection
   where Self::Position: Ord,
               Self::Whole: RandomAccessCollection,
{
}
```
Now with this anytime one needs to specify `R: RandomAccessCollection`, they need to explicitly specify `R::Position: Ord`. However, with the one I am talking about we don't need to use it. So, currently for Position, I used the first one and Whole I used the second one.

Personally I am not even sure why the first one even exists and what is the difference between 2 in terms of language level. Also not sure why constraints become this explicit with second syntax.

RishabhRD commented:
> I don't know why `prefix_mut` and `suffix_mut` need to consume their argument, maybe that's just a Rust style default for things that are likely to be small?

Actually its a rust style to pass small things by consuming but I don't like it personally and I prefer immutable reference only.

> In Hylo the rule of thumb is pass by value (immutable reference in Rust) unless the value is going to escape (e.g. be stored or returned).

I am actually storing positions inside `SliceMut` structure, that's why I passed it in a consuming way.
```rust
pub struct SliceMut<'a, Whole>
where
    Whole: ReorderableCollection<Whole = Whole>,
{
    whole: &'a mut Whole,
    from: Whole::Position,
    to: Whole::Position,
}
```
Do you think I should pass it by immutable reference only as `Position: Regular`?

RishabhRD commented:
> Would you like to discuss naming?

For sure, we should discuss naming and areas for improvement there.

> I don't really know how it's usually done in Rust, but I might have some useful input.

I got some tips from a rust expert. We might take care of that in discussion. Quoting their words as it:

```text
• Don’t name things MutableX, name them XMut instead. So MutableSlice
→ SliceMut, MutableArraySlice → ArraySliceMut.
• Don’t ever prefix field names with m_. So m_slice → slice, m_start →
start, m_end → end. The syntax of Rust makes this practice
unnecessary.
```

dabrahams commented:
> Current change is not an exact translation of the whole workaround. I took the idea of `Whole` thing only as I felt rust was able to handle the need for `CollectionCore` and `SliceCore`. I might miss something in this, please correct me if I am wrong here.

There are a couple of features my code supports that _maybe_ yours doesn't: 

1. A default slice type with the ability to customize it for any particular model of `Collection`.
2. The slice of a slice is exactly the original slice type, and not something like `Slice<Slice<X>>`.

I suspect you have no. 2 but might have missed no. 1.  If you have both features, you certainly don't need any of those types.  You might try fleshing out and adding [these demonstrations](https://gist.github.com/dabrahams/bc9497acdc20001cde14ed57bc4b4965#file-norecursiveconstraints2-swift-L118-L185) in some tests.

dabrahams commented:
I would implement the partition atop an underlying half-stable algorithm that partitions a collection on some value.

You could look at swift-algorithms for some inspiration (due to Swift limitations that should have been fixed years ago, they can't use a slice-first design).

> This is the specific syntax I am talking about:
> 
> ```rust
> pub trait RandomAccessCollection:
>     BidirectionalCollection<Position: Regular + Ord, Whole: RandomAccessCollection>
> {
> }
> ```
> 
> Rust supports it, just that it can't be recursive in nature like we have for Whole.
> 
> Ord constraint on Position would clarify this. Without this syntax it would be like:
> 
> ```rust
> pub trait RandomAccessCollection: BidirectionalCollection
>    where Self::Position: Ord,
>                Self::Whole: RandomAccessCollection,
> {
> }
> ```

I'm confused.  The difference between the two, on the surface, looks like a missing `Regular` constraint on the second one, not `Ord`.  Why should the first syntax behave differently from the second, apart from that?

> Now with this anytime one needs to specify `R: RandomAccessCollection`, they need to explicitly specify `R::Position: Ord`. However, with the one I am talking about we don't need to use it. So, currently for Position, I used the first one and Whole I used the second one.

I guess I don't understand this either.  I don't see those extra `Ord` constraints in your library code.

> Personally I am not even sure why the first one even exists and what is the difference between 2 in terms of language level. Also not sure why constraints become this explicit with second syntax.

OK, I guess we're both confused about this then.  Regardless, it may be possible to use some of the trait refactoring tricks in my prototype to avoid this recursive constraint limitation.

dabrahams commented:
> Do you think I should pass it by immutable reference only as `Position: Regular`?

Probably not.  The policy we anticipate for Hylo is that you would consume it because otherwise there may be a needless copy.

kyouko-taiga commented:
As I've been mentioned, I just wanted to write a small message to say that I've been following this discussion, that I find @RishabhRD 's work important and fascinating, and that I will be happy to contribute whatever insights I can bring to the table in the unlikely chance that I have some that Dave doesn't also have. Sadly I'm currently swamped with other things 😢

RishabhRD commented:
> 1. A default slice type with the ability to customize it for any particular model of `Collection`.

My design does not support this feature. In my understanding, in your workaround code Slice struct just forwards call to SliceCore and SliceCore is the customization point. Currently I am always returning Slice<Whole> where Slice is similar to DefaultSliceCore. 

I am afraid that it might not be possible to implement the slice core customization point this way because this customization point should again be a `lifetime generic associated type` that comes with same borrow checker limitations we were facing earlier.

I am curious about usecase of this customization point. The only usecase currently I can think of is proper bound safety for operations like after_n.

One of possible workaround I see for this depends on highly unstable nightly feature named `specialization`. The idea is if we have custom collection type C for which type author wants to provide custom slicing then they can provide specialization for collection conformance like `impl Collection for Slice<C>`.

RishabhRD commented:
>  I would implement the partition atop an underlying half-stable algorithm that partitions a collection on some value. You could look at swift-algorithms for some inspiration (due to Swift limitations that should have been fixed years ago, they can't use a slice-first design). 

Assuming you are talking about `partition(by:)` function here. I think we can't use `partition(by:)` function directly for quick sort purpose as it requires collection elements to be cloneable (and a cost of extra copy for elements like String) as immutable reference to pivot element can't be used as it would violate borrow checker independence rule. Thus an additional variant for partition algorithm would be needed like `partition_on_position` for quick sort implementation.

RishabhRD commented:
> Most of my thoughts about naming are orthogonal to your expert's advice.

Yep I am looking forward for your naming inputs. I feel there is lot to improve on this with current code (like after_n seems to be one).

> It's fascinating how my IDE turns function parameter names into things that aren't in the text but that look like argument labels at the call site, and how that drives one to name those parameters like argument labels rather than what they represent.

Same for me on Neovim with LSP.

RishabhRD commented:
Thanks, @kyouko-taiga! for your kind words :smiley:. I really appreciate that.

dabrahams commented:
> I am curious about usecase of this customization point. 

The use-cases are somewhat limited, but one category of example is collections that can be their own slice type, e.g. a memory buffer that stores a pointer and a length, or a range of integers. Otherwise, its Slice needs to store two more integers.

> The only usecase currently I can think of is proper bound safety for operations like after_n. Regarding after_n, I guess it can become more safe for Slice by adding after_n_limited_by kind of method.

I think I don't understand how `after_n` relates to this.

dabrahams commented:
@kyouko-taiga I think if not insights you have a unique perspective on lazy collections.  You've been suggesting YAGNI, that single-pass iterator-ish access is better.  I have more trouble making arguments for that position than you do.

RishabhRD commented:
> The use-cases are somewhat limited, but one category of example is collections that can be their own slice type, e.g. a memory buffer that stores a pointer and a length, or a range of integers. Otherwise, its Slice needs to store two more integers.

Got your point. With given use cases in my mind, I was trying if some careful use of unsafe constructs can make it possible and let me avoid lifetime GATs. Was not able to achieve much.

> I think I don't understand how after_n relates to this.

That was totally my overthinking :sweat_smile:. Now even I think about it, it is totally unrelated.

dabrahams commented:
Regarding laziness, I opened https://github.com/orgs/hylo-lang/discussions/1668

RishabhRD commented:
Merging this PR as positionless design is gonna take little time to explore and this is for sure a solid base for rs-stl purposes. Will open another PR for exploring laziness today.

I learnt a lot from this model and I also got some cool things to explore (laziness and position-less design). Thank You @dabrahams  for this whole thing, it was never possible without you tbh.
